cuda/test/factorization/par_ict_kernels.cpp

```diff
@@ -0,0 +1,177 @@
+        if (!input_file) {
+            FAIL() << "Could not find the file \"" << file_name
+                   << "\", which is required for this test.\n";
+        }
+        mtx_ani = gko::read<Csr>(input_file, ref);
+        mtx_ani->sort_by_column_index();
+
+        {
+            mtx_l_ani = Csr::create(ref, mtx_ani->get_size());
+            gko::matrix::CsrBuilder<value_type, index_type> l_builder(
+                lend(mtx_l_ani));
+            gko::kernels::reference::factorization::initialize_row_ptrs_l(
+                ref, lend(mtx_ani), mtx_l_ani->get_row_ptrs());
+            auto l_nnz =
+                mtx_l_ani->get_const_row_ptrs()[mtx_ani->get_size()[0]];
+            l_builder.get_col_idx_array().resize_and_reset(l_nnz);
+            l_builder.get_value_array().resize_and_reset(l_nnz);
+            gko::kernels::reference::factorization::initialize_l(
+                ref, lend(mtx_ani), lend(mtx_l_ani), true);
+        }
+        dmtx_ani->copy_from(lend(mtx_ani));
+        dmtx_l_ani->copy_from(lend(mtx_l_ani));
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::CudaExecutor> cuda;
+
+    const gko::dim<2> mtx_size;
+    std::default_random_engine rand_engine;
+
+    std::unique_ptr<Csr> mtx;
+    std::unique_ptr<Csr> mtx_ani;
+    std::unique_ptr<Csr> mtx_l_ani;
+    std::unique_ptr<Csr> mtx_l;
+
+    std::unique_ptr<Csr> dmtx;
+    std::unique_ptr<Csr> dmtx_ani;
+    std::unique_ptr<Csr> dmtx_l_ani;
+    std::unique_ptr<Csr> dmtx_l;
+};
+
+
+TEST_F(ParIct, KernelAddCandidatesIsEquivalentToRef)
+{
+    auto mtx_llt = Csr::create(ref, mtx_size);
+    mtx_l->apply(lend(mtx_l->transpose()), lend(mtx_llt));
+    auto dmtx_llt = Csr::create(cuda, mtx_size);
+    dmtx_llt->copy_from(lend(mtx_llt));
+    auto res_mtx_l = Csr::create(ref, mtx_size);
+    auto dres_mtx_l = Csr::create(cuda, mtx_size);
+
+    gko::kernels::reference::par_ict_factorization::add_candidates(
+        ref, lend(mtx_llt), lend(mtx), lend(mtx_l), lend(res_mtx_l));
+    gko::kernels::cuda::par_ict_factorization::add_candidates(
+        cuda, lend(dmtx_llt), lend(dmtx), lend(dmtx_l), lend(dres_mtx_l));
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(res_mtx_l, dres_mtx_l);
+    GKO_ASSERT_MTX_NEAR(res_mtx_l, dres_mtx_l, 1e-14);
+}
+
+
+TEST_F(ParIct, KernelComputeLUIsEquivalentToRef)
+{
+    auto square_size = mtx_ani->get_size();
+    auto mtx_l_coo = Coo::create(ref, square_size);
+    mtx_l_ani->convert_to(lend(mtx_l_coo));
+    auto dmtx_l_coo = Coo::create(cuda, square_size);
+    dmtx_l_coo->copy_from(lend(mtx_l_coo));
+
+    gko::kernels::reference::par_ict_factorization::compute_factor(
+        ref, lend(mtx_ani), lend(mtx_l_ani), lend(mtx_l_coo));
+    for (int i = 0; i < 20; ++i) {
+        gko::kernels::cuda::par_ict_factorization::compute_factor(
+            cuda, lend(dmtx_ani), lend(dmtx_l_ani), lend(dmtx_l_coo));
+    }
+
+    GKO_ASSERT_MTX_NEAR(mtx_l_ani, dmtx_l_ani, 1e-2);
```

Comment:
Isn't that a pretty high tolerance ? Or is it acceptable for the incomplete factorization ?

Reply:
We use the same for ParILU. The issue here is (like with a Jacobi iteration) that in the worst case (think tridiagonal matrix and updating elements front to back), we are only equivalent to the reference solution after n iterations.